src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -5214,9 +5222,10 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
       }
 
       def typedAlternative(alt: Alternative) = {
-        context withinPatAlternative (
-          treeCopy.Alternative(tree, alt.trees mapConserve (alt => typed(alt, mode, pt))) setType pt
-        )
+        context withinPatAlternative {
+          val trees1 = alt.trees mapConserve (alt => typed(alt, mode, pt))
+          treeCopy.Alternative(tree, trees1) setType lub(trees1 map (_.tpe))
+        }
```

Comment:
Corresponding code in dotty:

https://github.com/lampepfl/dotty/blob/972f6d5eff30539a5a8e4016ab2df3164adc76c3/compiler/src/dotty/tools/dotc/typer/Typer.scala#L1229-L1232

https://github.com/lampepfl/dotty/blob/972f6d5eff30539a5a8e4016ab2df3164adc76c3/compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala#L501-L502

```scala
  def typedAlternative(tree: untpd.Alternative, pt: Type)(implicit ctx: Context): Alternative = track("typedAlternative") {
    val trees1 = tree.trees mapconserve (typed(_, pt))
    assignType(cpy.Alternative(tree)(trees1), trees1)
  }

  def assignType(tree: untpd.Alternative, trees: List[Tree])(implicit ctx: Context) =
    tree.withType(ctx.typeComparer.lub(trees.tpes))
```